# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: security/manager/ssl/tests/unit/test_intermediate_preloads.js
# Commit: 889d7cd14e4d
# Full Hash: 889d7cd14e4d76462eabe4a26a551c72fb416e86
# Author: Dana Keeler <dkeeler@mozilla.com>
# Date: 2020-06-05 14:35:19
# Regressor Bug: 1630434
# File Overlap Count: 1
# Description:
#   Bug 1630434 - de-duplicate preloaded intermediates that may have been cached in cert9.db r=kjacobs,bbeurdouche
#   
#   In general, PSM caches intermediates from verified certificate chains in the
#   NSS certdb. Before bug 1619021, this would include preloaded intermediates,
#   which is unnecessary because cert_storage has a copy of those certificates, and
# ==============================================================================

diff -r d82b77ed1125 -r 889d7cd14e4d security/manager/ssl/tests/unit/test_intermediate_preloads.js
--- a/security/manager/ssl/tests/unit/test_intermediate_preloads.js	Fri Jun 05 00:39:30 2020 +0000
+++ b/security/manager/ssl/tests/unit/test_intermediate_preloads.js	Fri Jun 05 00:44:52 2020 +0000
@@ -586,6 +586,70 @@
   }
 );
 
+function findCertByCommonName(certDB, commonName) {
+  for (let cert of certDB.getCerts()) {
+    if (cert.commonName == commonName) {
+      return cert;
+    }
+  }
+  return null;
+}
+
+add_task(
+  {
+    skip_if: () => !AppConstants.MOZ_NEW_CERT_STORAGE,
+  },
+  async function test_healer() {
+    Services.prefs.setBoolPref(INTERMEDIATES_ENABLED_PREF, true);
+    Services.prefs.setIntPref(INTERMEDIATES_DL_PER_POLL_PREF, 100);
+
+    let certDB = Cc["@mozilla.org/security/x509certdb;1"].getService(
+      Ci.nsIX509CertDB
+    );
+    // Add an intermediate as if it had previously been cached.
+    addCertFromFile(certDB, "test_intermediate_preloads/int.pem", ",,");
+    // Add an intermediate with non-default trust settings as if it had been added by the user.
+    addCertFromFile(certDB, "test_intermediate_preloads/int2.pem", "CTu,,");
+
+    let syncResult = await syncAndDownload(["int.pem", "int2.pem"]);
+    equal(syncResult, "success", "Preloading update should have run");
+
+    equal(
+      (await locallyDownloaded()).length,
+      2,
+      "There should have been 2 downloads"
+    );
+
+    let healerRanPromise = TestUtils.topicObserved(
+      "psm:intermediate-preloading-healer-ran"
+    );
+    Services.prefs.setIntPref(
+      "security.intermediate_preloading_healer.timer_interval_ms",
+      500
+    );
+    Services.prefs.setBoolPref(
+      "security.intermediate_preloading_healer.enabled",
+      true
+    );
+    await healerRanPromise;
+    Services.prefs.setBoolPref(
+      "security.intermediate_preloading_healer.enabled",
+      false
+    );
+
+    let intermediate = findCertByCommonName(
+      certDB,
+      "intermediate-preloading-intermediate"
+    );
+    equal(intermediate, null, "should not find intermediate in NSS");
+    let intermediate2 = findCertByCommonName(
+      certDB,
+      "intermediate-preloading-intermediate2"
+    );
+    notEqual(intermediate2, null, "should find second intermediate in NSS");
+  }
+);
+
 function run_test() {
   server = new HttpServer();
   server.start(-1);
